libethash-cl/CLMiner.cpp

```diff
@@ -305,7 +306,7 @@ void CLMiner::workLoop()
 	try {
 		while (true)
 		{
-			const WorkPackage w = work();
+			const WorkPackage w = getWork();
```

Comment:
Why do you need this trick with "newwork" flag. Can't use the copy of work `w` here instead of asking for work with `work()` again?

Reply:
Not sure what you mean? That's in the open-cl miner. It's got to get its work somewhere, and this is the only place it does it. The problem in cuda was that the loop was using a work call when not really asking for work, but just needing to check its current content.

The new semantics are:

getWork() : if there is new work, make it current; return current work. Called once at the top of work loop of both miners.

work(): Give me the current work. Only used by cuda when evaluating found share in software.

Reply:
That's the problem: we should not use `work()` to evaluate the found share.

What data is missing to do the evaluation? Copy it to Solution so that the evaluation will not need to reach to `work()` again.

Reply:
Actually, I could change the names around such that there are no changes to CLMiner.

Rename the new getWork() to work(),

and call the current work() something else, like currentWork().

Reply:
Can you try my suggestion?

If not I can try it.

Reply:
CLMiner doesn't have this race problem because it calls report() in an orderly fashion directly from the work loop. CUDAMiner invokes report() asynchronously via directly from the .cu code.

As you pointed out, unnecessarily complicated, and slow! I'll be working on that.